pkg/tsdb/azuremonitor/azuremonitor.go

```diff
@@ -161,3 +164,185 @@ func (s *Service) newQueryMux() *datasource.QueryTypeMux {
+	request, err := http.NewRequest(http.MethodPost, url, bytes.NewBuffer(body))
+	request.Header.Set("Content-Type", "application/json")
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := dsInfo.Services["Azure Resource Graph"].HTTPClient.Do(request)
+	if err != nil {
+		return nil, err
+	}
+
+	return res, nil
+}
+
+func (s *Service) CheckHealth(ctx context.Context, req *backend.CheckHealthRequest) (*backend.CheckHealthResult, error) {
+	dsInfo, err := s.getDSInfo(req.PluginContext)
+	if err != nil {
+		return nil, err
+	}
+
+	status := backend.HealthStatusOk
+	metricsLog := "Successfully connected to Azure Monitor endpoint."
+	logAnalyticsLog := "Successfully connected to Azure Log Analytics endpoint."
+	graphLog := "Successfully connected to Azure Resource Graph endpoint."
+
+	metricsRes, err := checkAzureMonitorMetricsHealth(dsInfo)
+	if err != nil || metricsRes.StatusCode != 200 {
+		if err != nil {
+			return nil, err
+		} else {
+			body, err := io.ReadAll(metricsRes.Body)
+			if err != nil {
+				return nil, err
+			}
+			backend.Logger.Error(string(body))
+		}
+		status = backend.HealthStatusError
+		metricsLog = "Error connecting to Azure Monitor endpoint."
+	}
+
+	logsRes, err := checkAzureLogAnalyticsHealth(dsInfo)
+	if err != nil || logsRes.StatusCode != 200 {
+		status = backend.HealthStatusError
+		logAnalyticsLog = "Error connecting to Azure Log Analytics endpoint."
+		if err != nil {
+			if err.Error() == "no default workspace found" {
+				status = backend.HealthStatusUnknown
+				logAnalyticsLog = "No Log Analytics workspaces found."
+			} else {
+				return nil, err
+			}
+		} else {
+			body, err := io.ReadAll(logsRes.Body)
+			if err != nil {
+				return nil, err
+			}
+			backend.Logger.Error(string(body))
+		}
+	}
+
+	resourceGraphRes, err := checkAzureMonitorResourceGraphHealth(dsInfo)
+	if err != nil || resourceGraphRes.StatusCode != 200 {
+		if err != nil {
+			return nil, err
+		} else {
+			body, err := io.ReadAll(resourceGraphRes.Body)
+			if err != nil {
+				return nil, err
+			}
+			backend.Logger.Error(string(body))
+		}
+		status = backend.HealthStatusError
+		graphLog = "Error connecting to Azure Resource Graph endpoint."
+	}
+
+	defer func() {
+		if metricsRes != nil {
+			if err := metricsRes.Body.Close(); err != nil {
+				backend.Logger.Error("Failed to close response body", "err", err)
+			}
+		}
+		if logsRes != nil {
+			if err := logsRes.Body.Close(); logsRes != nil && err != nil {
+				backend.Logger.Error("Failed to close response body", "err", err)
+			}
+		}
+		if resourceGraphRes != nil {
+			if err := resourceGraphRes.Body.Close(); resourceGraphRes != nil && err != nil {
+				backend.Logger.Error("Failed to close response body", "err", err)
+			}
+		}
+	}()
+
+	return &backend.CheckHealthResult{
+		Status:  status,
+		Message: fmt.Sprintf("1. %s\n2. %s\n3. %s", metricsLog, logAnalyticsLog, graphLog),
```

Comment:
btw, are the `\n` being interpreted?

Reply:
They are not interpreted in the frontend

Reply:
Anyway, thinking about this, I think that the message is too repetitive (1. Successfully connected ... 2. Successfully connected... etc). We can probably simplify it. My suggestion is

- If the three checks pass, simply say "Successfully connected to Azure Monitor"
- If some check fails then we list each element. I noticed that you can specify "Details" when returning a message. This is helpful for this use case (and properly interprets `\n`). I tried:

```
	return &backend.CheckHealthResult{
		Status:  backend.HealthStatusError,
		Message: "Some health check failed",
		JSONDetails: []byte(
			`{"verboseMessage":"1. Successfuly connected to Azure Monitor Metrics\n2. Error connecting to Azure Resource Graph endpoint: Boom!"}`,
		),
	}, nil
```

and got:

![Screenshot from 2022-06-15 16-59-41](https://user-images.githubusercontent.com/4025665/173859864-0124214c-67ae-49b8-94bc-b235a0df678c.png)